### PR TITLE
FTP: Refactoring on SSH (jsch -> sshj) (previously "Secure FTP connection intermittently fails")

### LINKED ISSUE #365: Secure FTP connection intermittently fails
We use version 0.1.54 of jsch, which [appears to contain a bug](https://sourceforge.net/p/jsch/bugs/111/) that is triggered [since jdk 1.8.0_121](https://bugs.openjdk.java.net/browse/JDK-8174719).

We're impacted by this as shown by the test failures tracked in #361.

It looks like we'll either have to move away from jsch, wait for the bug to be fixed there, or help them fix it.

### PR DESCRIPTION
**Pull request by juanjovazquez**
The [sshj](https://github.com/hierynomus/sshj) library is used now instead of jsch for ssh handling. Both main and test code have been accordingly updated. The original and motivating issue regarding intermittent cryptography-related failures is no longer observed (at least for the time being).

closes #365

### CODE REVIEW

**Comment by juanjovazquez** [comment]
@argast According to #197 you worked on key based authentication for the sFTP integration. As explained in #365 we've had to change the ssh library we relied on. I've adapted the code but now the `StrictHostKeyCheckingSftpSourceSpec` test is not working. It seems that the client is trying to validate an RSA public key when the `known_hosts` file is declaring a DSA key. Could you please have it a look?. Thanks! :-)

**Comment by piotrkwiecinski** [comment]
@juanjovazquez in ```StrictHostKeyCheckingSftpSourceSpec``` when I removed 

```
.withOptions(new Pair("HostKeyAlgorithms", "+ssh-dss"))
```
Tests pass.

In ```OpenSSHKnownHosts``` in sshj there is detection of key types:

https://github.com/hierynomus/sshj/blob/master/src/main/java/net/schmizz/sshj/transport/verification/OpenSSHKnownHosts.java#L183

Maybe it's enough? Bare in mind my knowledge is quite limited :)

**Comment by argast** [comment]
Yes, this option was added because test key was in less secure des format which required additional ssh option to authenticate with it.

**Comment by piotrkwiecinski** [comment]
@juanjovazquez actually I've made mistake I have the same error...I'll try to play around with it tomorrow.

File: ftp/src/main/scala/akka/stream/alpakka/ftp/impl/SftpOperations.scala [hunk 1]
```diff
@@ -4,92 +4,112 @@
...
+  def connect(connectionSettings: SftpSettings)(implicit ssh: SSHClient): Try[Handler] = Try {
+    import connectionSettings._
 
-  private def configureIdentity(sftpIdentity: SftpIdentity)(implicit ftpClient: JSch) = sftpIdentity match {
-    case identity: RawKeySftpIdentity =>
-      ftpClient.addIdentity(identity.name, identity.privateKey, identity.publicKey.orNull, identity.password.orNull)
-    case identity: KeyFileSftpIdentity =>
-      ftpClient.addIdentity(identity.privateKey, identity.publicKey.orNull, identity.password.orNull)
-  }
+    // ssh connection
```
Comment (svezfaz) [inline ftp/src/main/scala/akka/stream/alpakka/ftp/impl/SftpOperations.scala]: minor: maybe we could do without these comments

File: ftp/src/main/scala/akka/stream/alpakka/ftp/impl/SftpOperations.scala [hunk 2]
```diff
@@ -4,92 +4,112 @@
...
+  }
 
-  def listFiles(handler: Handler): immutable.Seq[FtpFile] = listFiles(".", handler) // TODO
+  def listFiles(handler: Handler): immutable.Seq[FtpFile] = listFiles(".", handler)
 
   def retrieveFileInputStream(name: String, handler: Handler): Try[InputStream] = Try {
-    handler.get(name)
+    val remoteFile = handler.open(name, Set(OpenMode.READ).asJava)
+    val is = new remoteFile.RemoteFileInputStream()
+    if (is != null) is else throw new IOException(s"$name: No such file or directory")
```
Comment (svezfaz) [inline ftp/src/main/scala/akka/stream/alpakka/ftp/impl/SftpOperations.scala]: could be 
```
Option(is).getOrElse(throw new IOException(s"$name: No such file or directory"))
```

File: ftp/src/main/scala/akka/stream/alpakka/ftp/impl/SftpOperations.scala [hunk 3]
```diff
@@ -4,92 +4,112 @@
...
   }
 
   def storeFileOutputStream(name: String, handler: Handler, append: Boolean): Try[OutputStream] = Try {
-    handler.put(name, if (append) APPEND else OVERWRITE)
+    import OpenMode._
+    val openModes = Set(WRITE, CREAT) ++ (if (append) Set(APPEND) else Set())
+    val remoteFile =
+      handler.open(name, openModes.asJava)
+    val os = new remoteFile.RemoteFileOutputStream()
+    if (os != null) os else throw new IOException(s"Could not write to $name")
```
Comment (svezfaz) [inline ftp/src/main/scala/akka/stream/alpakka/ftp/impl/SftpOperations.scala]: as above

File: ftp/src/main/scala/akka/stream/alpakka/ftp/impl/SftpOperations.scala [hunk 4]
```diff
@@ -4,92 +4,112 @@
...
+    val is = new remoteFile.RemoteFileInputStream()
+    if (is != null) is else throw new IOException(s"$name: No such file or directory")
   }
 
   def storeFileOutputStream(name: String, handler: Handler, append: Boolean): Try[OutputStream] = Try {
-    handler.put(name, if (append) APPEND else OVERWRITE)
+    import OpenMode._
+    val openModes = Set(WRITE, CREAT) ++ (if (append) Set(APPEND) else Set())
+    val remoteFile =
+      handler.open(name, openModes.asJava)
```
Comment (svezfaz) [inline ftp/src/main/scala/akka/stream/alpakka/ftp/impl/SftpOperations.scala]: formatting: this could be put on the line above

**Comment by juanjovazquez** [comment]
@svezfaz Your requests have been applied. Thanks!.

**Comment by raboof** [comment]
What is the status of this PR? Is this now a complete replacement for the problematic approach or is there still work to do around host checking?

**Comment by juanjovazquez** [comment]
@raboof IMHO this PR is ready for merging.

**Pull request merged**